tests/sxg_sig_test.cc

```diff
@@ -26,13 +28,65 @@
 
 namespace {
 
-void FillSignature(sxg_buffer_t* buffer) {
+// Splices sig value out of buffer and into signature.
+void ExtractSignature(sxg_buffer_t* buffer, sxg_buffer_t* signature) {
   static const char sigKey[] = "sig=";
-  char* point = strstr((char*)buffer->data, sigKey) + sizeof(sigKey) - 1;
-  EXPECT_EQ('*', *point++);
-  while (*point != '*') {
-    *point++ = '%';
-  }
+  uint8_t* begin =
+      (uint8_t*)strstr((char*)buffer->data, sigKey) + sizeof(sigKey);
+  EXPECT_EQ('*', *(begin - 1));
+
+  uint8_t* end = begin;
+  while (*++end != '*')
+    ;
+  EXPECT_EQ('*', *end);
+
+  EXPECT_TRUE(sxg_buffer_resize(end - begin, signature));
+  memcpy(signature->data, begin, end - begin);
+
+  memmove(begin, end, (buffer->data + buffer->size) - end);
+  EXPECT_TRUE(sxg_buffer_resize(buffer->size - (end - begin), buffer));
+}
+
+// On failure, this leaks a EVP_ENCODE_CTX. Do not use this in production or in
+// a loop in test.
+void sxg_base64decode(const sxg_buffer_t* src, sxg_buffer_t* dst) {
+  const size_t offset = dst->size;
+  // 4-byte blocks to 3-byte, assuming none are padding chars; we'll adjust at
+  // the end.
+  const EVP_ENCODE_BLOCK_T estimated_out_length = 3 * (src->size / 4);
+
+  // EVP_DecodeBlock doesn't support padding chars, so we use the long form.
+  ASSERT_TRUE(sxg_buffer_resize(offset + estimated_out_length, dst));
+  EVP_ENCODE_CTX* ctx = EVP_ENCODE_CTX_new();
```

Comment:
as mentioned in comments, `EVP_ENCODE_CTX_new` is deprecated in BoringSSL. why can't we just use `EVP_EncodeBlock` directly?

Reply:
I took another look at this, and I see what you mean regarding padding. When I was looking earlier, I was focused on the encode side. That said, I think we can strip the padding bytes manually. Something like this:
```
void sxg_base64decode(const sxg_buffer_t* src, sxg_buffer_t* dst) {
  const size_t offset = dst->size;
  // 4-byte blocks to 3-byte, allowing for padding chars;
  const EVP_ENCODE_BLOCK_T estimated_out_length = 3 * (src->size / 4);

  ASSERT_TRUE(sxg_buffer_resize(offset + estimated_out_length, dst));
  ASSERT_EQ(EVP_DecodeBlock(dst->data + offset, src->data, src->size), estimated_out_length);

  // EVP_DecodeBlock ignores padding, so we have to strip null bytes manually
  EVP_ENCODE_BLOCK_T actual_out_length = estimated_out_length;
  while(dst->data[actual_out_length-1] == 0) {
    actual_out_length--;
  }
  
  ASSERT_TRUE(sxg_buffer_resize(offset + actual_out_length, dst));
}
```
works for me with both OpenSSL and BoringSSL.

Reply:
I think this might fail in the very rare case that the signature ends in a NUL byte? IIRC, that's how integer zero is encoded in DER.

We could count the padding chars at the end ourselves and know how many to strip. If my latest attempt doesn't work, I'll try that approach.